### ISSUE TITLE
Feature Request: Multiple Shadow Masks per definition *.txt file.

### ISSUE THREAD

**Issue by tonurics** [MiSTer-devel/ShadowMasks_MiSTer#47]
I would like to request an update to the Shadow Mask format to allow multiple masks within the same mask *.txt definition file. Along with conditional logic checks for selecting which Shadow Mask should be used. The goal being: a reduction of Shadow Mask files end-users have to choose from.

I've come up with three conditional checks below. Any one of them will make good progress towards the goal. Ideally, all three checks could implemented; but as two of them require end-user input/GUI options: I won't get my hopes up for a full implemention. 

For conditional logic purposes: I would suggest that the first set of conditions that satisfies requirements would be the Mask used and the parser stops reading Masks at that point. If no conditions are met: then the last Mask the file is used. If we are able to implement more than one of these checks, then: `*` should be added as a wildcard to test True for the check it's used on.

---
### Screen Height

This option requires no end-user input and can happen automatically for them. The check would be on the minimum screen height.

For example, here's a definition *.txt file for three Shadow Masks intended for 1440p, 1080p and everything else:

```
v3

1,6,1440

72a
72a
72a
02a
02a
02a

1,4,1080

72a
72a
02a
02a

1,2

72a
02a
```

The last entry could also be written as:
```
1,2,720

72a
02a
```
Even though it sets the minimum screen height to 720: as the last entry is still serves as a catch all.

The parsing logic would need to ensure that the conditional is rechecked in the event of a resolution change.

---
### Sub-Pixel Layout

This check would require the end-user to select the sub-pixel layout that works best for their screen via a GUI setting. Which the parser would then use for conditional checks.

Ignoring rotation [as end-users already have control over that via GUI] the six possible sub-pixel layouts are: BGR, BRG, GBR, GRB, RBG, RGB

RGB and BGR are the most common layouts [I would suggest RGB as the default]. But the other layouts do appear in the wild regularly: https://www.rtings.com/tv/tools/table/73403

Please give extra consideration to this specific check: I think the sub-pixel layouts are the trickiest aspect of Shadow Masks that end-users have to think about. The GUI setting would be great for hiding the issue to any end-users who are: none the wiser or won't need know about it [because they have a screen with a RGB layout]. Leaving only those end-users who see something wrong, to then: use the setting to find the layout for their screens. This check would reduce the amount of mask files, dumb things down for everyone, and preserve advanced options for users who would need them.

Building on the conditional logic above, allowing `|` as a deliminator for multiple options in this check would be ideal. [The actual parser could just perform a string search; with any character other than R, G, or B serving as a deliminator.]

An example combining the resolution conditional check above with this one could look like:

```
v3

4,6,1080,BGR|BRG|GBR

72a,12a,22a,42a
72a,12a,22a,42a
72a,12a,22a,42a
22a,42a,72a,12a
22a,42a,72a,12a
22a,42a,72a,12a

4,6,1440,RGB

72a,42a,22a,12a
72a,42a,22a,12a
72a,42a,22a,12a
22a,12a,72a,42a
22a,12a,72a,42a
22a,12a,72a,42a

4,4,1080,BGR|BRG|GBR

72a,12a,22a,42a
72a,12a,22a,42a
22a,42a,72a,12a
22a,42a,72a,12a

4,4,1080,RGB|RBG|GRB

72a,42a,22a,12a
72a,42a,22a,12a
22a,12a,72a,42a
22a,12a,72a,42a

4,2,*,BGR|BRG

72a,12a,22a,42a
22a,42a,72a,12a

4,2,*,GBR

72a,22a,12a,42a
12a,42a,72a,22a

4,2

72a,42a,22a,12a
22a,12a,72a,42a

```
The parsing logic would need to ensure that the conditional is rechecked in the event the user changes their sub-pixel layout setting.

---
### Bright / Dark

This check would require the end-user to select a Bright or Dark via a boolean GUI toggle [0 for Dark and 1 for Bright]. Which the parser would then use for conditional checks.

Many masks variations are created using RGB/CMY or K (Black)/W (White) as a way of providing brightness choices to users [not related to intensity levels of the mask themselves]. This is especially true in CRT masks that use Black or White "filler" between the colours. It would ultimately fall to the Mask creator to decide what "Bright" and "Dark" actually mean; i.e. the Dark version of a Mask might use CMYK whereas the Bright version of the same mask might use RGBW [or even a different pattern altogether].

Perhaps, the GUI element of this check can be combined with one for the Sub-Pixel Layout check. Again stressing how important I think that Sub-Pixel Layout check is to dumbing everything down for end-users.

An example combining all conditional checks could look like:

```
v3

6,4,1080,GBR,0

22a,12a,42a,22a,12a,42a
22a,12a,42a,22a,12a,42a
22a,12a,42a,02a,02a,02a
22a,12a,42a,22a,12a,42a
22a,12a,42a,22a,12a,42a
02a,02a,02a,22a,12a,42a

6,4,*,*,0

52a,62a,32a,52a,62a,32a
52a,62a,32a,02a,02a,02a
52a,62a,32a,52a,62a,32a
02a,02a,02a,52a,62a,32a

6,4

42a,22a,12a,42a,22a,12a
42a,22a,12a,72a,72a,72a
42a,22a,12a,42a,22a,12a
72a,72a,72a,42a,22a,12a
```
The parsing logic would need to ensure that the conditional is rechecked in the event the user changes their preference.

**Comment by tonurics** [comment]
Here's a quick breakdown of the reductions each check could provide to their sets:
| Condition | Current | After | Reduction |
| --- | --- | --- | --- |
| Screen Height | 3* | 1 | -2 |
| Sub-Pixel Layout | 6 | 1 | -5 |
| Bright / Dark | 2 | 1 | -1 |

_*Assuming masks with sizes at: 720p, 1080p & 1440p_

Combining "Sub-Pixel Layout" with "Bright / Dark" would allow 12 PenTile masks to reduce into a single file.

The "Sub-Pixel Layout" check provides the best over all benefit, as any mask using colour could need a variation for each end-user layout. Whereas the "Screen Height" and "Bright / Dark " apply to a smaller subset of masks that have those variations.

---

An alternative, and perhaps better solution for the "Sub-Pixel Layout" problem could be: to only create masks for the RGB layout and programmatically change the colours based on the end-user's setting.

**Comment by sorgelig** [comment]
I've added muti-resolution shadow masks. It will check just vertical resolution.
There will be multiple masks in one file. each mask will start with resolution=X string which means any vertical resolution higher or equal to X will use following mask.
mask for highest resolution lower than current resolution will be used
if current resolution is lower than lowest listed in mask file, then first mask will be used.

That will be pretty much final version. Sub-pixel layout and dark/bright modes is a matter of specific shadow mask file. There is no point to put different masks into one file.

Again, don't make simple things complicated. 99% users will need basically up to 3-4 masks only. Very complex shadow mask definition will most likely limit to only you who will be able to make it. Other masks will remain simple and OSD option won't work as expected (or always will be unavailable).

I will make release probably tomorrow. Too much hassle for a very niche feature!

**Comment by sorgelig** [comment]
This repo will be re-organized later. It will have several (5-6) most common masks in the root. All other masks will be in sub-folder "others". This should simplify the choice for generic user.

**Comment by tonurics** [comment]
I'm happy to hear about the multiple resolutions. I suspected that conditional had the best chance of being implemented.

I'd like to get started on updating the Editor, so it's ready for your changes. And had a couple of questions towards that end: 
- Will you follow the Mask ordering logic I described above? [Where the order of the Mask within the file is important.]
- Is `resolution=X` on it's own new line, or appended to the "width,height" line used by each mask?

Perhaps you can just share a simple example *.txt file with a few masks.

---

As for making things overly complicated: there have been many differing opinions as to how we should simplify choices for end-users. This feature request was the result of my efforts to provide editorial guidance to all of those suggestions and concerns; along with solutions to those ideas which I think could be achieved.

This request is not born of my own dissatisfaction. I am just trying to be helpful and guide feedback into meaningful directions.

**Comment by sorgelig** [comment]
you can check this:
[Resolution test.txt](https://github.com/MiSTer-devel/ShadowMasks_MiSTer/files/7808681/Resolution.test.txt)


> * Will you follow the Mask ordering logic I described above? [Where the order of the Mask within the file is important.]

not sure what ordering you mean. It's basically several masks concatenated together with resolution tags

**Comment by tonurics** [comment]
> mask for highest resolution lower than current resolution will be used

I miss read that line in your original comment. So never mind. ;)

For clarity: I suggested above, we don't keep track of each Mask; instead we simply use the first Mask [in the file] that meets the conditions [with the last Mask being used if no conditions are met]. So in the case of Height: the Masks would need to appear in order of largest to smallest.

That would make the file parsing logic a lot simpler; as you wouldn't need to read all the Masks, then find the largest one that is also less than the current screen Height.

**Comment by sorgelig** [comment]
it's not hard to parse. But you always can list resolutions from lowest to highest.
First mask eventually doesn't need "resolution" tag but still can have it.

**Comment by tonurics** [comment]
Yeah, `Resolution=0` would be implied for anything that doesn't have it set. Which is what I was going to default the GUI Editor to set.

I assume you will bump the version to `v3` the example *.txt file you shared above is still saying `v2`.

**Comment by sorgelig** [comment]
it's still v2 because changes are not confusing the parser.

**Comment by tonurics** [comment]
This request has been partially fulfilled. With the remaining items described as "WontFix".
